### PR TITLE
[docs] Migrate Developer update for 4.0

### PR DESCRIPTION
**Pull request by sarjona**
Migrated pages:
- https://docs.moodle.org/dev/Moodle_4.0_developer_update

<a href="https://gitpod.io/#https://github.com/moodle/devdocs/pull/184"></a>

### CODE REVIEW

File: versioned_docs/version-4.0/devupdate.md [hunk 1]
```diff
@@ -0,0 +1,958 @@
...
+
+:::note
+
+This is currently only possible on an activity and block level.
+
+:::
+
+### Tertiary navigation
+
+The action buttons have been moved to the top of the page and would encourage you to do the same.
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:52]:
```suggestion
Action buttons have been moved to the top of the page and we would encourage you to do the same.
```

File: versioned_docs/version-4.0/devupdate.md [hunk 2]
```diff
@@ -0,0 +1,958 @@
...
+If you have any buttons on an activity page that go to another page, or open a form (or similar, then we encourage you to move them from the body of your activity page to the top. All of the core activities have been updated to follow this pattern. Please take a look to see how you can format your activity in a similar fashion. There is no API here. You are welcome to create the buttons and display them as you wish in this top area.
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:53]:
```suggestion
If you have any buttons on an activity page that go to another page, or open a form (or similar), then we encourage you to move them from the body of your activity page to the top. All of the core activities have been updated to follow this pattern. Please take a look to see how you can format your activity in a similar fashion. There is no API here. You are welcome to create the buttons and display them as you wish in this top area.
```

File: versioned_docs/version-4.0/devupdate.md [hunk 3]
```diff
@@ -0,0 +1,958 @@
...
+
+- To leverage the activity_header, it's data needs to be exported and included into the base template :
+  - `headercontent` being the array element that contains the exported activity_header data
+
+  ```html
+  {{#headercontent}}
+     {{> core/activity_header}}
+  {{#headercontent}}
+
+   ```
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:129]:
```suggestion
  {{#headercontent}}
   ```
```

File: versioned_docs/version-4.0/devupdate.md [hunk 4]
```diff
@@ -0,0 +1,958 @@
...
+Each Moodle installation now ships with a Moodle User Interface (UI) Component library, a documentation system used to describe all the Bootstrap components and the custom Moodle components. The component Library is a helper tool for developers when creating user interfaces, a testing tool for theme developers and a documentation tool for core developers. The ultimate goal of having a component library is to encourage developers to create consistent user interfaces to improve Moodle's overall user experience.
+
+The library contains pages with documentation about User Interface components. It contains details on how to use the component, what variations are available and the JavaScript events / options are associated with the component.
+
+When writing on these pages it is possible to render core mustache templates using some custom syntax like this:
+
+```mustache
+ {{< mustache template="core/notification_error" >}}
+ {{< /mustache >}}
+```
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:168]:
```suggestion
```mustache
{{< mustache template="core/notification_error" >}}
{{< /mustache >}}
```
```

File: versioned_docs/version-4.0/devupdate.md [hunk 5]
```diff
@@ -0,0 +1,958 @@
...
+
+You can also call core JavaScript or use HTML examples where the HTML code and the rendered result are visible in the Component Library. For more info visit the [Moodle templates](http://componentlibrary.moodle.com/admin/tool/componentlibrary/docspage.php/library/moodle-templates/) page or the [Moodle JavaScript](http://componentlibrary.moodle.com/admin/tool/componentlibrary/docspage.php/library/moodle-javascript/) page.
+
+Each page in the library uses the current CSS from the default theme in your Moodle installation, if you have multiple themes installed and enabled the setting "Allow theme changes on url", the component library will have a theme selector option.
+
+:::tip
+
+A hosted version of the Component Library can be found here. http://componentlibrary.moodle.com
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:176]:
```suggestion
A hosted version of the Component Library can be found at [http://componentlibrary.moodle.com](http://componentlibrary.moodle.com).
```

File: versioned_docs/version-4.0/devupdate.md [hunk 6]
```diff
@@ -0,0 +1,958 @@
...
+### Enabling the Component Library
+
+Component library pages are written in the markdown language. These pages need to be compiled to HTML pages before the Component Library is visible. To compile the pages the server running Moodle needs to have [nodeJS and Grunt tools](/general/development/tools/nodejs).
+
+If your server meets all requirements you can enable the library running
+
+```sh
+ $ npm install
+ $ grunt componentlibrary
+```
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:189]:
```suggestion
```sh
$ npm install
$ grunt componentlibrary
```
```

File: versioned_docs/version-4.0/devupdate.md [hunk 7]
```diff
@@ -0,0 +1,958 @@
...
+- New features should be build keeping in mind the UI part needs to be customisable and if possible (and making sense) reusable. And example would be the new page drawers that we are introducing for the Navigation project. Or the custom primary navigation menus where overflowing items are pushed into a More section.
+
+## Theme changes
+
+### Edit switch
+
+On theme boost the "Turn editing on" and "Customise this page" buttons have been replaced by an edit switch in the top navbar. Theme Classic will keep using the old buttons. Child themes can choose to use the edit switch if the theme config.php is using this variable
+
+```php
+ $THEME->haseditswitch = true;
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:209]:
```suggestion
$THEME->haseditswitch = true;
```

File: versioned_docs/version-4.0/devupdate.md [hunk 8]
```diff
@@ -0,0 +1,958 @@
...
+
+For example:
+
+```gherkin
+Given 100 "mod_lti > tool types" exist with the following data:
+  |name        |Test tool [](count)                  |
+  |description |Example description [](count)        |
+  |baseurl     |https://www.example.com/tool[](count)|
+
+```
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:368]:
```suggestion
```gherkin
Given 100 "mod_lti > tool types" exist with the following data:
  | name        | Test tool [](count)                   |
  | description | Example description [](count)         |
  | baseurl     | https://www.example.com/tool[](count) |

```
```

File: versioned_docs/version-4.0/devupdate.md [hunk 9]
```diff
@@ -0,0 +1,958 @@
...
+```gherkin
+And I click on the "tab title" dynamic tab
+```
+
+To make sure that edit mode is (or is not) available on the current page, the following steps can be used.
+
+```gherkin
+And edit mode should be available on the current page
+And edit mode should not be available on the current page
+
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:381]:
```suggestion
```

File: versioned_docs/version-4.0/devupdate.md [hunk 10]
```diff
@@ -0,0 +1,958 @@
...
+```gherkin
+And I switch editing mode on
+And I switch editing mode off
+```
+
+In addition to the existing step to go to a course with editing mode on, we now have the following step to do the same but with editing mode being off.
+
+```gherkin
+And I am on "course full name" course homepage with editing mode off
+
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:395]:
```suggestion
```

File: versioned_docs/version-4.0/devupdate.md [hunk 11]
```diff
@@ -0,0 +1,958 @@
...
+And I enable "course_summary" "block" plugin
+```
+
+::::
+
+The following is a special variation of `I click on "<page name>" "link" in the "page" "region"`. It first checks to see if we are on the given page via the breadcrumb. If not we then attempt to follow the link name given.
+
+```gherkin
+And I follow the breadcrumb "page name"
+
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:434]:
```suggestion
```

File: versioned_docs/version-4.0/devupdate.md [hunk 12]
```diff
@@ -0,0 +1,958 @@
...
+```gherkin
+And I should see "Node" is active in navigation
+```
+
+:::info For example:
+
+```gherkin
+And I should see "My courses" is active in navigation
+
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:447]:
```suggestion
```

File: versioned_docs/version-4.0/devupdate.md [hunk 13]
```diff
@@ -0,0 +1,958 @@
...
+```gherkin
+And I select "Node" from primary navigation
+```
+
+The following steps are to check whether an item exists in the user menu or not.
+
+```gherkin
+And "item text" "selector type" should exist in the user menu
+And "item text" "selector type" should not exist in the user menu
+
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:463]:
```suggestion
```

File: versioned_docs/version-4.0/devupdate.md [hunk 14]
```diff
@@ -0,0 +1,958 @@
...
+:::
+
+The following steps are to check if a submenu of the user menu is shown, and if an item exists or does not exist in a given user submenu.
+
+```gherkin
+And I should see "submenu name" user submenu
+And "item text" "selector type" should exist in the "submenu name" user submenu
+And "item text" "selector type" should not exist in the "submenu name" user submenu
+
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:480]:
```suggestion
```

File: versioned_docs/version-4.0/devupdate.md [hunk 15]
```diff
@@ -0,0 +1,958 @@
...
+And "English (pirate) &lrm;(en_ar)&lrm;" "link" should not exist in the "Language selector" user submenu
+```
+
+:::
+
+To add enrolment methods to courses you can use the following new step. The data that you provide in the next lines are used to fill the enrolment method form.
+
+```gherkin
+And I add "enrolment method" in "course identifier" with:
+
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:498]:
```suggestion
```

File: versioned_docs/version-4.0/devupdate.md [hunk 16]
```diff
@@ -0,0 +1,958 @@
...
+:::
+
+There are 3 new steps specific to the calendar component. These steps can be used to hover over a day in the mini-calendar or the full calendar.
+
+```gherkin
+And I hover over day "day of month" of this month in the full calendar page
+And I hover over day "day of month" of this month in the mini-calendar block
+And I hover over today in the mini-calendar block
+
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:518]:
```suggestion
```

File: versioned_docs/version-4.0/devupdate.md [hunk 17]
```diff
@@ -0,0 +1,958 @@
...
+Some new steps are added to question bank to be able to add comment to questions, verify the existence of a comment, and deleting comments form questions. Please note that the steps for adding comments only write the comment text in the comment field. You still need to click on the "Add comment" button to save the comment.
+
+```gherkin
+And I add "comment text" comment to question
+And I add "comment text" comment to question preview
+And I delete "comment text" comment from question
+And I delete "comment text" comment from question preview
+And I should see "number of comments" on the comments column
+And I click "number of comments" on the row on the comments column
+
```
Comment (andrewnicols) [inline versioned_docs/version-4.0/devupdate.md:530]:
```suggestion
```

**Review by andrewnicols — CHANGES_REQUESTED** [review]
Mostly minor changes. There's a lot of whitespace surrounding all of the code blocks! I stopped commenting them all.

File: versioned_docs/version-4.0/devupdate.md (hunk 4 above)
Comment (sarjona) [reply versioned_docs/version-4.0/devupdate.md:168]: I don't think these blockquotes should be closed...

File: versioned_docs/version-4.0/devupdate.md (hunk 6 above)
Comment (sarjona) [reply versioned_docs/version-4.0/devupdate.md:189]: I don't think these blockquotes should be removed.

File: versioned_docs/version-4.0/devupdate.md (hunk 8 above)
Comment (sarjona) [reply versioned_docs/version-4.0/devupdate.md:368]: I will remove the empty line but will keep the closing blockquote ;-)

Comment (andrewnicols) [reply versioned_docs/version-4.0/devupdate.md:366]: Good plan!

File: versioned_docs/version-4.0/devupdate.md (hunk 6 above)
Comment (andrewnicols) [reply versioned_docs/version-4.0/devupdate.md:187]: No - that was unintentional. I suspect the suggestion is confused by the amount of backticks

**Comment by andrewnicols** [comment]
As discussed, this currently needs to be in /docs and not the versioned_docs. This is just for now, and next release we'll be ahead of the curve and this won't be an issue.

Also, this could do with some tags.

File: versioned_docs/version-4.0/devupdate.md [hunk 18]
```diff
@@ -0,0 +1,958 @@
...
+The primary navigation (the navbar) apart from the existing content will now display links to the Dashboard, My Courses, Site Admin and Course search, by default. You can still add items to the navbar via the `custom menu` option. This will be displayed within the **'More'** menu.
+
+:::note
+
+The menus have been transitioned to be rendered via templates. Refer `user_menu.mustache`. The lang menu has been moved to reside within the user menu.
+
+:::
+
+#### Customising the primary navigation
```
[the remark this answers is not in the document]
Comment (sarjona) [reply versioned_docs/version-4.0/devupdate.md:21]: Good idea! Done!

**Comment by sarjona** [comment]
Hi @andrewnicols! Thanks for looking at the page! I've reviewed all the raised points, so the PR is ready again to get your eyes on it! ;-)

File: docs/devupdate.md [hunk 19]
```diff
@@ -0,0 +1,958 @@
+---
+title: Moodle 4.0 developer update
+tags: []
```
Comment (andrewnicols) [inline docs/devupdate.md:3]: Could still benefit from some tags I think here

**Review by andrewnicols — CHANGES_REQUESTED** [review]
Hi Sara,

Just two minor things. Feel free to self-merge once solved as I'm heading off for the night.

File: versioned_sidebars/version-4.0-sidebars.json [hunk 20]
```diff
@@ -1,5 +1,11 @@
 {
```
[the remark this answers is not in the document]
Comment (sarjona) [reply versioned_sidebars/version-4.0-sidebars.json]: You're right! :-) Done!

File: docs/devupdate.md (hunk 19 above)
Comment (sarjona) [reply docs/devupdate.md]: My apologies! I totally forget it!! I've added a couple of tags to this page! :-) Well spotted!

**Pull request merged**